Documentation

konradkonrad opened the pull request:
This adds some documentation for SDK usage. Human friendly preview: https://github.com/shutter-network/shop-sdk/tree/doc?tab=readme-ov-file#shutter-shop-sdk

fredo commented on README.md:
```diff
@@ -1,15 +1,314 @@
+Shutter-Optimism is an [Optimism](https://www.optimism.io/) testnet, that provides a mechanism to
+include Shutter encrypted transactions.
+
+[Shutter](http://shutter.network/) is a distributed key generation (DKG) threshold encryption
```
Would put shutter explainer in its own section

fredo commented on README.md:
```diff
@@ -1,15 +1,314 @@
+by not releasing decryption keys, and _collusion_, by peeking into encrypted messages before the
+reveal condition was reached, become very unlikely.
+
+In Shutter-Optimism this allows to build
```
This sentence is somewhat odd to me. I would try to highlight that the underlying chain allows for MEV-resistant usage when transactions are shutterized. 

It should not be confused that the contracts are inherently MEV resistant or that they would need specific adaptions to be MEV resistant.

fredo commented on README.md:
```diff
@@ -1,15 +1,314 @@
+In Shutter-Optimism this allows to build
+[MEV](https://ethereum.org/en/developers/docs/mev/)-resistant applications.
+
+The intention of `shop-sdk` is to make it as easy as possible to transform a traditional
```
ah should have read the next sentence ...

fredo commented on README.md:
```diff
@@ -1,15 +1,314 @@
+
+[Decryption](https://pkg.go.dev/github.com/shutter-network/shutter/shlib@v0.1.18/shcrypto#EncryptedMessage.Decrypt)
+of messages encrypted with the above scheme requires an `epoch secret key`, that is created
+collaboratively by the keyper set, based on a certain `epoch identifier`.
```
I'm missing the explainer that the epoch corresponds to the block number

fredo commented on README.md:
```diff
@@ -1,15 +1,314 @@
+  // protocol that allows Ethers access to make all read-only
+  // requests through MetaMask.
+  // SHUTTER> use ShutterProvider
+  provider = new ShutterProvider(options, window.ethereum)
```
Question: Is it possible to create a ShutterProvider instance by using an RPC url?

konradkonrad replied:
I don't think so - `ShutterProvider` [extends `BrowserProvider`](https://github.com/ethers-io/ethers.js/blob/v6.12.1/src.ts/providers/provider-browser.ts#L44), so you'd need to wrap your RPC url in an implementation of an [`Eip1193Provider`](https://github.com/ethers-io/ethers.js/blob/v6.12.1/src.ts/providers/provider-browser.ts#L16) (usually provided by the browser plugin). I guess it is potentially doable, but I don't know how much effort it would be.

fredo commented on README.md:
````diff
@@ -1,15 +1,314 @@
+  await signer._sendTransactionTrace(txRequest, inclusionWindow)
+```
+
+#### Inspecting shutterized blocks
````
Is this explanation valid and usable if there were multiple encrypted transactions in a block or is this specific to a single encrypted transaction for this epoch?

konradkonrad replied:
yes, this should trace all `executions` (note the plural) of encrypted transactions

fredo replied:
Ok what would it take to filter logs for a specific transaction? I think that's what people want to have

fredo commented:
nitpick: In darkmode, the graphic is basically invisible

konradkonrad commented:
> I think it's a good first version.

Thanks!

> As a developer who has no experience how shutter works I would still be quite confused how the actual transaction flow works and differs from normal execution. I would dedicate a whole section to that with pictures and graphs. It will probably save us a lot of time if the user actually knows how transactions are flowing before getting into how to use the SDK. I'm happy to split the work and have another PR for this.

I've added a simple flow graph, to help developing a mental model.

Not sure, if I would go into much more detail here: I've added plenty of links to research the underlying technologies. IMO we should equip a developer with enough to get started building and avoid duplicating information, that may stay more up to date elsewhere.

fredo commented:
> > I think it's a good first version.
> 
> Thanks!
> 
> > As a developer who has no experience how shutter works I would still be quite confused how the actual transaction flow works and differs from normal execution. I would dedicate a whole section to that with pictures and graphs. It will probably save us a lot of time if the user actually knows how transactions are flowing before getting into how to use the SDK. I'm happy to split the work and have another PR for this.
> 
> I've added a simple flow graph, to help developing a mental model.
> 
> Not sure, if I would go into much more detail here: I've added plenty of links to research the underlying technologies. IMO we should equip a developer with enough to get started building and avoid duplicating information, that may stay more up to date elsewhere.

👍 Flowgraph should be sufficient. They jsut should be informed about the OP stack implementation specifics and what concerns them. Things like "In contrast of the original transaction flow, users will send the encrypted transaction as data into an inbox contract, from where the sequencer will them pick up to execute later. Encrypted transactions are scheduled to be executed in chosen blocks. This results in what we call inclusion tx (normal tx, writing encrypted data into the inbox and scheduling it for tx in block X) and the execution transaction (happens on scheduled block and will always be the first tx in the block)
Not saying that these information are not already in there but we need to make sure that this concept is understood by the devs.

konradkonrad commented:
@fredo Feel free to merge, unless you have further comments!

fredo commented:
Thanks for addressing all comments! Will merge and we'll take it from there.
